src/apify/_utils.py

```diff
@@ -410,3 +412,118 @@ def validate_single(field_value: Any, expected_type: type, required: bool, name:
 PARSE_DATE_FIELDS_MAX_DEPTH = 3
 PARSE_DATE_FIELDS_KEY_SUFFIX = 'At'
 ListOrDictOrAny = TypeVar('ListOrDictOrAny', list, dict, Any)
+
+
+def get_short_base64_hash(data: bytes) -> str:
+    """Generates an 8-character, base64-encoded SHA-256 hash of the provided data.
+
+    Args:
+        data: The binary data to be hashed.
+
+    Returns:
+        An 8-character, base64-encoded hash.
+    """
+    # Create a SHA-256 hash of the payload.
+    hash_object = sha256(data)
+    # Convert the hash to a base64-encoded string.
+    base64_encoded_hash = b64encode(hash_object.digest()).decode()
+    # Remove any '+' or '/' or '=' characters and return the first 8 characters.
+    return base64_encoded_hash.replace('+', '').replace('/', '').replace('=', '')[:8]
+
+
+def normalize_url(url: str, *, keep_url_fragment: bool = False) -> str:
+    """Normalizes a URL.
+
+    This function cleans and standardizes a URL by removing leading and trailing whitespaces,
+    converting the scheme and netloc to lower case, stripping unwanted tracking parameters
+    (specifically those beginning with 'utm_'), sorting the remaining query parameters alphabetically,
+    and optionally retaining the URL fragment. The goal is to ensure that URLs that are functionally
+    identical but differ in trivial ways (such as parameter order or casing) are treated as the same.
+
+    Args:
+        url: The URL to be normalized.
+        keep_url_fragment: Flag to determine whether the fragment part of the URL should be retained.
+
+    Returns:
+        A string containing the normalized URL. If normalization fails, the original URL is returned. It logs
+            a warning message in this case.
+    """
+    try:
+        # Parse the URL
+        parsed_url = urlparse(url.strip())
+        search_params = dict(parse_qsl(parsed_url.query))  # Convert query to a dict
+
+        # Remove any 'utm_' parameters
+        search_params = {k: v for k, v in search_params.items() if not k.startswith('utm_')}
+
+        # Construct the new query string
+        sorted_keys = sorted(search_params.keys())
+        sorted_query = urlencode([(k, search_params[k]) for k in sorted_keys])
+
+        # Construct the final URL
+        new_url = (
+            parsed_url._replace(
+                query=sorted_query,
+                scheme=parsed_url.scheme,
+                netloc=parsed_url.netloc,
+                path=parsed_url.path.rstrip('/'),
+            )
+            .geturl()
+            .lower()
+        )
+
+        # Retain the URL fragment if required
+        if not keep_url_fragment:
+            new_url = new_url.split('#')[0]
+
+    except Exception as exc:
+        logger.warning(f'Failed to normalize URL: {exc}')
+        return url
```

Comment:
I'd rather make the caller responsible for handling the exception - functions like `normalize_url` probably shouldn't write logs (side effects).